packages/overlay/src/OverlayBase.ts

```diff
@@ -0,0 +1,891 @@
+import {
+    ifDefined,
+    styleMap,
+} from '@spectrum-web-components/base/src/directives.js';
+import styles from './overlay-base.css.js';
+import { overlayStack } from './OverlayStack.js';
+import { PlacementController } from './PlacementController.js';
+import { OverlayTypes } from './overlay-types.js';
+import { OverlayTimer } from './overlay-timer.js';
+
+export const overlayTimer = new OverlayTimer();
+
+export type OpenableElement = HTMLElement & {
+    open: boolean;
+    tipElement?: HTMLElement;
+    updateComplete?: Promise<void>;
+};
+
+const LONGPRESS_DURATION = 300;
+
+type LongpressEvent = {
+    source: 'pointer' | 'keyboard';
+};
+
+export const LONGPRESS_INSTRUCTIONS = {
+    touch: 'Double tap and long press for additional options',
+    keyboard: 'Press Space or Alt+Down Arrow for additional options',
+    mouse: 'Click and hold for additional options',
+};
+
+export class BeforetoggleClosedEvent extends Event {
+    currentState = 'open';
+    newState = 'closed';
+    constructor() {
+        super('beforetoggle', {
+            bubbles: false,
+            composed: false,
+        });
+    }
+}
+
+export class BeforetoggleOpenEvent extends Event {
+    currentState = 'closed';
+    newState = 'open';
+    constructor() {
+        super('beforetoggle', {
+            bubbles: false,
+            composed: false,
+        });
+    }
+}
+
+export const noop = (): void => {
+    return;
+};
+
+/**
+ * Apply a "transitionend" listener to an element that may not transition but
+ * guarantee the callback will be fired either way.
+ *
+ * @param el {HTMLElement} - Target of the "transition" listeners.
+ * @param action {Function} - Method to trigger the "transition".
+ * @param cb {Function} - Callback to trigger when the "transition" has ended.
+ */
+export const guaranteedTransitionend = (
+    el: HTMLElement,
+    action: () => void,
+    cb: () => void
+): void => {
+    const cleanup = (): void => {
+        el.removeEventListener('transitionrun', handleTransitionrun);
+        el.removeEventListener('transitionend', handleTransitionend);
+        cb();
+    };
+    let guarantee2: number;
+    let guarantee3: number;
+    // WebKit fires `transitionrun` a little earlier, so the inner/outer relationship
+    // here allows WebKit to be caught, but doesn't remove the animation listener until
+    // after it would have fired in Chromium.
+    const guarantee1 = requestAnimationFrame(() => {
+        guarantee2 = requestAnimationFrame(() => {
+            guarantee3 = requestAnimationFrame(() => {
+                cleanup();
+            });
+        });
+    });
+    const handleTransitionend = (event: TransitionEvent): void => {
+        if (event.propertyName === 'visibility') {
+            // Ignore "visibility" transitions because they often happen before/after a
+            // larger transition and don't represent the overall transition duration.
+            return;
```

Comment:
Should we document this exception, or are visibility-only transitions rare enough to assume they don't happen?

Reply:
Often what happens in this:
```
    transition: transform 0.5s ease-in-out,
        opacity 0.5s ease-in-out,
        visibility 0s linear;
```
The goal is to go from "invisible" to "visible" _immediately_, while transitioning other values at a slower rate. At it is matched with the following:
```
    transition: transform 0.5s ease-in-out,
        opacity 0.5s ease-in-out,
        visibility 0s linear 0.5s;
```
Where the goal is to transition over values at a slower rate and then at the very edit of the transition switch form "visible" to "invisible" _immediately_.

The other path I thought of here was to count up the `transitionrun` events, and then count down the `transitionend` events, so we could know 100% that _ALL_ of the transitions had completed. Let me know if you think that would be better here than the `visibility` caveats that currently are leveraged.

Reply:
> The other path I thought of here was to count up the transitionrun events, and then count down the transitionend events, so we could know 100% that ALL of the transitions had completed. Let me know if you think that would be better here than the visibility caveats that currently are leveraged.

❤️  perfect

Reply:
Please checkout the additional docs to go with this.